plugins/in_head/in_head.c

```diff
@@ -0,0 +1,256 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2015-2016 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+
+#include <msgpack.h>
+#include <fluent-bit/flb_input.h>
+#include <fluent-bit/flb_config.h>
+#include <fluent-bit/flb_error.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_stats.h>
+
+#include "in_head.h"
+
+/* cb_collect callback */
+static int in_head_collect(struct flb_config *config, void *in_context)
+{
+    struct flb_in_head_config *head_config = in_context;
+    int     fd        = -1;
+    ssize_t read_size = 0;
+    int     ret = -1;
+
+    /* open at every collect callback */
+    fd = open(head_config->filepath, O_RDONLY);
+    if ( fd < 0) {
+        perror("open");
+        goto collect_fin;
+    }
+
+    head_config->buf_len = read(fd, head_config->buf,
+                                head_config->buf_size);
+    flb_debug("%s read_len=%d buf_size=%d",__FUNCTION__,
+              head_config->buf_len, head_config->buf_size);
+
+    if (head_config->buf_len < 0){
+        perror("read");
+        goto collect_fin;
+    }
+    head_config->buf[head_config->buf_len -1] = '\0';
```

Comment:
On this line, you are setting a NULL character in the wrong position, e.g:

   {"head":"96996.47 187893.47\u0000"}

two workarounds:

- right position: head_config->buf[head_config->buf_len] = '\0';
- since you are packaging directly to msgpack, there is no need for this.

so I think number 2 is better.